src/OrleansTestKit/TestGrainCreator.cs

```diff
@@ -42,21 +47,27 @@ public Grain CreateGrainInstance<T>(IGrainContext context)
             throw new ArgumentNullException(nameof(context));
         }
 
+        // this needs to remain set for the lifetime of the grain -- Orleans has hooks to validate that its set
+        SetExecutionContextMethod(context);
```

Comment:
`RuntimeContext.SetExecutionContext` sets the private field [`_threadLocalContext`](https://github.com/dotnet/orleans/blob/v7.1.2/src/Orleans.Core.Abstractions/Runtime/RuntimeContext.cs#L11-L15) which is `[ThreadStatic]`. AFAIK, there is no guarantee or option available to ensure that an xunit test runner will run the setup, SUT, and teardown on the same thread. In fact, this short xunit discussion seems to indicate that it's explicitly _not_ possible to enforce it in the VS runner:

https://github.com/xunit/xunit/discussions/2167

I know that we have test fixtures in our project built on `IAsyncLifetime`, and the placement of the `CreateGrainInstance` call relative to the `await`s could easily translate into this breaking. Check out this demo, notably the change in output after the await on line 35:

https://gist.github.com/JoshKeegan/66d649f6c5cc127d2f4057133d7d5bdc

Assume a scenario without asynchronous setup code, though. Can we even guarantee that an xunit test runner doesn't change threads between synchronous setup code and the SUT? I'm not sure it's possible to track down all test runner implementations and provide this guarantee.

This all works fine when running in an Orleans runtime thanks to the use of the custom task scheduler [that calls `SetExecutionContext`](https://github.com/dotnet/orleans/blob/v7.1.2/src/Orleans.Runtime/Scheduler/ActivationTaskScheduler.cs#L41) before running the C# `Task` that represents the grain method call. However, the runtime also handles continuations (or "turns") to ensure that `RuntimeContext.Current` is consistent before and after any `await`s.

There isn't an easy answer to this issue, and I'm concerned that this change will not address anything but the simplest of use cases. It's possible that's the intent... but I'm skeptical of changes that lead to incongruent experiences. I'd instead prefer that the TestKit fail fast and explicitly warn the developer that this scenario is unsupported. (I might look into that- how good are the errors when `RuntimeContext.Current` are `null`?)

Reply:
Responded to your comments in the main thread.